hpl-2.2/man/man3/HPL_indxg2l.3
.TH HPL_indxg2l 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_indxg2l \- Map a global index into a local one.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&int\fR
\fB\&HPL_indxg2l(\fR
\fB\&const int\fR
\fI\&IG\fR,
\fB\&const int\fR
\fI\&INB\fR,
\fB\&const int\fR
\fI\&NB\fR,
\fB\&const int\fR
\fI\&SRCPROC\fR,
\fB\&const int\fR
\fI\&NPROCS\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_indxg2l\fR
computes  the local index of a matrix entry pointed to by
the  global index IG.  This  local  returned index is the same in all
processes.
.SH ARGUMENTS
.TP 8
IG      (input)                 const int
On entry, IG specifies the global index of the matrix  entry.
IG must be at least zero.
.TP 8
INB     (input)                 const int
On entry,  INB  specifies  the size of the first block of the
global matrix. INB must be at least one.
.TP 8
NB      (input)                 const int
On entry,  NB specifies the blocking factor used to partition
and distribute the matrix. NB must be larger than one.
.TP 8
SRCPROC (input)                 const int
On entry, if SRCPROC = -1, the data  is not  distributed  but
replicated,  in  which  case  this  routine returns IG in all
processes. Otherwise, the value of SRCPROC is ignored.
.TP 8
NPROCS  (input)                 const int
On entry,  NPROCS  specifies the total number of process rows
or columns over which the matrix is distributed.  NPROCS must
be at least one.
.SH SEE ALSO
.BR HPL_indxg2lp \ (3),
.BR HPL_indxg2p \ (3),
.BR HPL_indxl2g \ (3),
.BR HPL_numroc \ (3),
.BR HPL_numrocI \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pauxil/HPL_indxg2l.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
int HPL_indxg2l
(
   const int                        IG,
   const int                        INB,
   const int                        NB,
   const int                        SRCPROC,
   const int                        NPROCS
)
#else
int HPL_indxg2l
( IG, INB, NB, SRCPROC, NPROCS )
   const int                        IG;
   const int                        INB;
   const int                        NB;
   const int                        SRCPROC;
   const int                        NPROCS;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_indxg2l computes  the local index of a matrix entry pointed to by
 * the  global index IG.  This  local  returned index is the same in all
 * processes.
 *
 * Arguments
 * =========
 *
 * IG      (input)                       const int
 *         On entry, IG specifies the global index of the matrix  entry.
 *         IG must be at least zero.
 *
 * INB     (input)                       const int
 *         On entry,  INB  specifies  the size of the first block of the
 *         global matrix. INB must be at least one.
 *
 * NB      (input)                       const int
 *         On entry,  NB specifies the blocking factor used to partition
 *         and distribute the matrix. NB must be larger than one.
 *
 * SRCPROC (input)                       const int
 *         On entry, if SRCPROC = -1, the data  is not  distributed  but
 *         replicated,  in  which  case  this  routine returns IG in all
 *         processes. Otherwise, the value of SRCPROC is ignored.
 *
 * NPROCS  (input)                       const int
 *         On entry,  NPROCS  specifies the total number of process rows
 *         or columns over which the matrix is distributed.  NPROCS must
 *         be at least one.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   int                        i, j;
/* ..
 * .. Executable Statements ..
 */
   if( ( IG < INB ) || ( SRCPROC == -1 ) || ( NPROCS == 1 ) )
/*
 * IG  belongs  to the first block,  or the data is not distributed,  or
 * there is just one process in this dimension of the grid.
 */
      return( IG );
/*
 * IG  =  INB - NB + ( l * NPROCS + MYROC ) * NB + X  with  0 <= X < NB,
 * thus IG is to be found in the block (IG-INB+NB) / NB = l*NPROCS+MYROC
 * with  0 <= MYROC < NPROCS.  The local index to be returned depends on 
 * whether  IG  resides in the process owning the first partial block of
 * size INB (MYROC=0). To determine this cheaply, let i = (IG-INB) / NB,
 * so that if NPROCS divides i+1, i.e. MYROC=0,  we have i+1 = l*NPROCS.
 * If we set  j = i / NPROCS, it follows that j = l-1. Therefore, i+1 is
 * equal to (j+1) * NPROCS.  Conversely, if NPROCS does not divide  i+1, 
 * then i+1 = l*NPROCS + MYROC with 1 <= MYROC < NPROCS. It follows that
 * j=l and thus (j+1)*NPROCS > i+1.
 */
   j = ( i = ( IG - INB ) / NB ) / NPROCS;
/*
 * When IG resides in the process owning the first partial block of size
 * INB (MYROC = 0), then the result IL can be written as:
 * IL = INB - NB + l * NB + X  = IG + ( l - (l * NPROCS + MYROC) ) * NB.
 * Using the above notation,  we have i+1 = l*NPROCS + MYROC = l*NPROCS,
 * i.e l = ( i+1 ) / NPROCS = j+1,  since  NPROCS divides i+1, therefore
 * IL = IG + ( j + 1 - ( i + 1 ) ) * NB.
 *
 * Otherwise when MYROC >= 1, the result IL can be written as:
 * IL = l * NB + X = IG - INB + ( ( l+1 ) - ( l * NPROCS + MYROC ) )*NB.
 * We still have i+1 = l*NPROCS+MYROC. Since NPROCS does not divide i+1,
 * we have j = (l*NPROCS+MYROC-1) / NPROCS = l, i.e
 * IL = IG - INB + ( j + 1 - ( i + 1 ) ) * NB.
 */
   return( NB * (j - i) + 
           ( ( i + 1 - ( j + 1 )*NPROCS ) ? IG - INB : IG ) );
/*
 * End of HPL_indxg2l
 */
}
